File: examples/complete/07-sieve-bench/go/bench.go
// 07-sieve-bench (Go): same workload as the Spectra twin.
// N=500 (95 primes), iters=200, checksum total = 19000.
package main

import (
	"fmt"
	"os"
)

func main() {
	const n = 500
	const iters = 200
	total := 0
	for it := 0; it < iters; it++ {
		sieve := make([]int, n+1)
		for p := 2; p*p <= n; p++ {
			if sieve[p] == 0 {
				for m := p * p; m <= n; m += p {
					sieve[m] = 1
				}
			}
		}
		count := 0
		for k := 2; k <= n; k++ {
			if sieve[k] == 0 {
				count++
			}
		}
		if count != 95 {
			fmt.Fprintln(os.Stderr, "prime count mismatch")
			os.Exit(1)
		}
		total += count
	}
	if total != 19000 {
		fmt.Fprintln(os.Stderr, "checksum mismatch")
		os.Exit(1)
	}
	fmt.Println("SIEVE ok primes=95 total=19000")
}

File: examples/complete/07-sieve-bench/rust/bench.rs
// 07-sieve-bench (Rust): same workload as the Spectra twin.
// N=500 (95 primes), iters=200, checksum total = 19000.

fn main() {
    const N: usize = 500;
    const ITERS: i64 = 200;
    let mut total: i64 = 0;
    for _ in 0..ITERS {
        let mut sieve = vec![0i64; N + 1];
        let mut p: usize = 2;
        while p * p <= N {
            if sieve[p] == 0 {
                let mut m = p * p;
                while m <= N {
                    sieve[m] = 1;
                    m += p;
                }
            }
            p += 1;
        }
        let mut count: i64 = 0;
        for k in 2..=N {
            if sieve[k] == 0 {
                count += 1;
            }
        }
        if count != 95 {
            eprintln!("prime count mismatch");
            std::process::exit(1);
        }
        total += count;
    }
    if total != 19000 {
        eprintln!("checksum mismatch");
        std::process::exit(1);
    }
    println!("SIEVE ok primes=95 total=19000");
}
